Set OwnerReferences on helm release configmaps after install/upgrade

chancez opened the pull request:
The OwnerReference is set to the CR that triggered the install/upgrade,
meaning if the CR is deleted, the helm release configMaps are also
deleted.

Before this completely works, we'll need to also set ownerReferences of the resources created from helm, and remove the logic for running `helm delete`, as the release configMap will be deleted when the CR is, and helm will be unable to find the release when it attempts to run helm delete.

Commit:
Set OwnerReferences on helm release configmaps after install/upgrade

The OwnerReference is set to the CR that triggered the install/upgrade,
meaning if the CR is deleted, the helm release configMaps are also
deleted.

chancez commented:
Still need to test in minikube, but I put the PR up so that I can get some input on some questions I had. See the todo here: https://github.com/coreos/lostromos/pull/4/files#diff-5c1cbb789cf325597283901eb3fcf5deR127

In my helm-operator MVP I ended up setting the ownerReferences before install and after, to ensure that if the helm install/upgrade never succeeds the ownerReferences for the configmaps are still set. In my MVP I also set the ownerReferences at startup in case of a crash or something. Do we think these are necessary?

ericavonb commented on helmctlr/helm.go:
```diff
@@ -151,6 +169,46 @@ func (c Controller) marshallCR(r *unstructured.Unstructured) ([]byte, error) {
+	rlsName := c.releaseName(r)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"OWNER": "TILLER",
```
Make "TILLER" a const set at top?

ecordell commented on helmctlr/helm.go:
```diff
@@ -122,23 +124,39 @@ func (c Controller) installOrUpdate(r *unstructured.Unstructured) error {
 	}
 
 	rlsName := c.releaseName(r)
+	// TODO: should we set configmap owner references before install/update, in
```
Doesn't that depend on when the configmap gets created? Do we know how to query for it before the release is made?

to be safe, I would think we want to do:

```
- check if configmap exists
   - if so, make sure it has ownerrefs
- helm update/install
- check if configmap exists
   - if so, make sure it has ownerrefs
```

chancez replied:
We can basically query the same way both times, it's pretty easy, we would be effectively listing configmaps before we run the release to see if any previous release CMs never got their owner refs set (ie, in case we fail multiple times in a row during/after install/upgrade).

chancez replied:
I've updated to have the code do the check/update before, and after install/upgrade.

ecordell commented on helmctlr/helm.go:
```diff
@@ -151,6 +169,46 @@ func (c Controller) marshallCR(r *unstructured.Unstructured) ([]byte, error) {
+			continue
+		}
+		controllerRef := metav1.NewControllerRef(r, r.GroupVersionKind())
+		ownerRefs := []metav1.OwnerReference{
```
do we need to be careful here? shouldn't we pull existing ownerrefs from the configmap first and only add to them (unless it's to the same CR group/version/kind, then overwrite?)

chancez replied:
Tiller currently doesn't set owner references, so I'm pretty sure we're fine, but I could try to support that.

chancez replied:
Since we know tiller isn't currently using ownerRefs configmaps, and we're going to be prescribing how to run tiller/what version, can we punt on this?

ecordell commented on helmctlr/helm_test.go:
```diff
@@ -137,6 +160,28 @@ func TestResourceAddedHappyPath(t *testing.T) {
 		testController.ResourceAdded(testResource)
 	})
+
+	// Ensure helm configMap owner references get set
```
This seems fine unless we need to handle additional states (see previous comments)

chancez replied:
Okay, I don't think we really need to handle any other states really, this should be sufficient for now.

ecordell commented on helmctlr/helm_test.go:
```diff
@@ -124,6 +143,10 @@ func TestResourceAddedHappyPath(t *testing.T) {
 	testController.Helm = mockHelm
+
+	k8sClient := fake.NewSimpleClientset(testHelmReleaseConfigmap)
+	testController.KubeClient = k8sClient
```
I would rather see testController take the kubernetes client as an input instead of making KubeClient public on the controller type

chancez replied:
Ah yes, I wasn't able to do that before cause I was using a generated mock rather than the fake client. Part of this all is because the current tests use a global testController.

chancez replied:
Fixed.

Commit:
helmctlr: Make "TILLER" label value const

Commit:
helmctlr: Set CM ownerReferences before install/upgrade also

Commit:
helmctlr: Pass fakeK8sClient to NewController in tests

No need to export kubeClient anymore since the fake can be created as a
global and passed to the constructor.

chancez commented:
@ecordell @ericavonb Updated the PR based on the feedback.
